File: src/main/cpp/connectivity.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>
#include <sstream>
#include <stack>

std::string find_connectivity();

using namespace std;

int main() {
    cout << find_connectivity();
    return 0;
}

std::string find_connectivity() {
    int vertexes, edges;
    cin >> vertexes >> edges;
    unordered_map<int, vector<int> > graph;
    for (int i = 0; i < edges; i++) {
        int from, to;
        cin >> from >> to;
        vector<int> adjacent;
        if (graph.find(from) != graph.end()) {
            graph.at(from).push_back(to);
        } else {
            adjacent.push_back(to);
            graph.insert(pair<int, vector<int> >(from, adjacent));
        }

        vector<int> inversed_adjacent;
        if (graph.find(to) != graph.end()) {
            graph.at(to).push_back(from);
        } else {
            inversed_adjacent.push_back(from);
            graph.insert(pair<int, vector<int> >(to, inversed_adjacent));
        }
    }
    string builder;
    int connectivityCount = 0;
    bool visited[vertexes + 1];
    for (int i = 0; i < vertexes + 1; ++i) {
        visited[i] = false;
    }
    for (int i = 1; i <= vertexes; i++) {
        if (visited[i]) {
            continue;
        }
        std::stack<int> stack;
        stack.push(i);

        int max_idx = -1;
        bool conn_bit_map[vertexes];
        for (int j = 0; j < vertexes; j++) {
            conn_bit_map[j] = false;
        }
        while (!stack.empty()) {
            int cur_node = stack.top();
            stack.pop();
            if (visited[cur_node]) {
                continue;
            } else {
                visited[cur_node] = true;
            }
            max_idx = max(max_idx, cur_node);
            conn_bit_map[cur_node - 1] = true;

            vector<int> adjacent;
            if (graph.find(cur_node) != graph.end()) {
                for (int vertex: graph.at(cur_node)) {
                    stack.push(vertex);
                }
            }
        }
        for (int j = 1; j <= max_idx; j++) {
            if (conn_bit_map[j - 1]) {
                builder += to_string(j) + " ";
            }
        }
        builder += "\n";
        connectivityCount++;
    }
    return to_string(connectivityCount) + "\n" + builder;
}

File: src/main/java/graph/Connectivity.java
package graph;

import java.io.*;
import java.util.*;

public class Connectivity {
    public static void main(String[] args) throws IOException {
        System.out.println(wrap(System.in));
    }

    public static String wrap(InputStream is) throws IOException {
        try (final BufferedReader reader = new BufferedReader(new InputStreamReader(is))) {
            final String[] firstLine = reader.readLine().split(" ");
            final int vertexes = Integer.parseInt(firstLine[0]);
            final int edges = Integer.parseInt(firstLine[1]);
            final Map<Integer, List<Integer>> graph = new HashMap<>(vertexes + 700);

            for (int i = 0; i < edges; i++) {
                final Integer[] edge = Arrays.stream(reader.readLine().split(" "))
                        .map(Integer::parseInt)
                        .toArray(Integer[]::new);
                final List<Integer> adjacent = graph.getOrDefault(edge[0], new ArrayList<>());
                adjacent.add(edge[1]);
                graph.put(edge[0], adjacent);

                final List<Integer> adjacentInversed = graph.getOrDefault(edge[1], new ArrayList<>());
                adjacentInversed.add(edge[0]);
                graph.put(edge[1], adjacentInversed);
            }

            final StringBuilder builder = new StringBuilder();
            int connectivityCount = 0;
            final boolean[] visited = new boolean[vertexes + 1];
            for (int j = 1; j <= vertexes; j++) {
                if (visited[j]) {
                    continue;
                }
                final Deque<Integer> stack = new LinkedList<>();
                stack.push(j);

                int maxIdx = -1;
                final boolean[] connectivityLst = new boolean[vertexes];
                while (!stack.isEmpty()) {
                    final Integer curNode = stack.pop();
                    maxIdx = Math.max(maxIdx, curNode);
                    if (visited[curNode]) {
                        continue;
                    } else {
                        visited[curNode] = true;
                    }
                    connectivityLst[curNode - 1] = true;
                    final List<Integer> adjacent = graph.get(curNode);
                    if (adjacent == null) {
                        continue;
                    }
                    adjacent.forEach(stack::push);
                }
                for (int i = 1; i <= maxIdx; i++) {
                    if (connectivityLst[i - 1]) {
                        builder.append(i).append(" ");
                    }
                }
                builder.append("\n");

                connectivityCount++;
            }
            return builder
                    .insert(0, "\n")
                    .insert(0, connectivityCount)
                    .toString();
        }
    }
}
